### PR TITLE
Proposed fix for GC rooting + extensive notes on how/why are in TimerToken

### PR DESCRIPTION
**Pull request by mgravell**
Adds automatic root/unroot; please read notes in the `TimerToken` class (in `ConnectionMultiplexer.cs`)

### CODE REVIEW

**Comment by mgravell** [comment]
Test failure is expecting `RedisTimeoutException` but getting `RedisConnectionException` with "The message timed out in the backlog attemtping to send because no connection became available" - not sure if this is my stuff or the stuff earlier in your branch, @NickCraver, but... I'm almost thinking we could legitimately change that RCE to a RTE; it *is a timeout*

File: src/StackExchange.Redis/ServerEndPoint.cs [hunk 1]
```diff
@@ -990,5 +990,12 @@ internal void SimulateConnectionFailure(SimulatedFailureType failureType)
             interactive?.SimulateConnectionFailure(failureType);
             subscription?.SimulateConnectionFailure(failureType);
         }
+
+        internal bool HasPendingCallerFacingItems()
+        {
+            // check whichever bridges exist
+            if (interactive is { } tmp && tmp.HasPendingCallerFacingItems()) return true;
```
Comment (NickCraver) [inline src/StackExchange.Redis/ServerEndPoint.cs:997]:
```suggestion
            if (interactive?.HasPendingCallerFacingItems() == true) return true;
```

**Comment by mgravell** [comment]
one test still failing (`AbortOnConnectFailTests.DisconnectAndReconnectThrowsConnectionExceptionSync`) - message has changed to:

> UnableToConnect on 127.0.0.1:6379/Interactive, Initializing/NotStarted, last: NONE, origin: BeginConnectAsync, outstanding: 0, last-read: 0s ago, last-write: 0s ago, keep-alive: 100s, state: Connecting, mgr: 0 of 10 available, last-heartbeat: never, global: 0s ago, v: 2.6.113.56944

@NickCraver  you know more about this test than me; care to opine? You mentioned that this was one of the brittle ones; we seem to have accidentally made it non-brittle by making it reliably fail; should we handle both messages? or is this wrong?

File: docs/ReleaseNotes.md [hunk 2]
```diff
@@ -8,7 +8,7 @@ Current package versions:
 
 ## Unreleased
 
-No pending changes.
+- Fix [#2392](https://github.com/StackExchange/StackExchange.Redis/issues/2392): critical (but rare) GC bug that can lead to async tasks never completing if the multiplexer is not held by the consumer ([#2413 by mgravell](https://github.com/StackExchange/StackExchange.Redis/pull/2413))
```
[the remark this answers is not in the document]
Comment (NickCraver) [reply docs/ReleaseNotes.md:11]: +1 - we'll refine release notes here, want to test with all combined manually

**Comment by NickCraver** [comment]
@mgravell +1 on test changed - I think that's fine, the goal of that specific test set isn't the type thrown (that's checked very explicitly in the exception tests), so being more lenient there is perfectly fine by me.

**Comment by NickCraver** [comment]
Status: my only concern here (hopefully unfounded) is lock perf - we're going to run some tests on that front and then should be good to merge in - I'll update release docs and all that goodness here but not imperative in this PR since it's going into the other.

**Comment by mgravell** [comment]
Oh, right; forgot I wasn't merging to main! My PR number will be confusing - we should fix that in the final outer PR

**Pull request merged**